Speed up binary_tidy: 37min -> 8min

pamarcos opened the pull request:
Linking takes a lot of the time for `binary_tidy`. Building also takes a bit, but since we're using `sccache` it's barely noticeable. In any case, we don't need to produce real binaries during the analysis stage, at least on the CI.

This PR speeds up the `binary_tidy` step 4.5x: from [37min](https://s3.amazonaws.com/clickhouse-test-reports/0/e71acb5494847769543b6b00fdc468244e9b4549/builds/report.html) to [8min](https://s3.amazonaws.com/clickhouse-test-reports/73647/fbfcfc8aafebc61d55329905f687ae3164f50a6e/builds/report.html):

- Use dummy launchers by default for tidy builds. The launcher is used for building, linking and stripping. The launcher just creates an empty file to trick the build system. The tricky part is that we still need to use a real compiler for some parts of the code that generate other source files required: protoc, llvm-tlbgen and all their dependants. We surround those CMakeLists.txt with the prolog `set_original_launchers_if_needed` and close it with `set_dummy_launchers_if_needed`.
- Allow running the build along with the static analysis by setting `ENABLE_DUMMY_LAUNCHERS=0`
- Add the `binary_tidy` build for sync PRs. This job was broken for several days. Let's prevent that by running it now that it's very fast
- Build job testing that the check works as expected after introducing a linting error on purpose at [`4b90e75` (#73647)](https://github.com/ClickHouse/ClickHouse/pull/73647/commits/4b90e75b4d5c20a7bc58bee0fc84a32900b58a73): https://github.com/ClickHouse/ClickHouse/actions/runs/12430017196/job/34710092969?pr=73647

### Changelog category (leave one):
- CI Fix or Improvement (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ] <!---ci_set_required--> Allow: All Required Checks
- [ ] <!---ci_include_stateless--> Allow: Stateless tests
- [ ] <!---ci_include_stateful--> Allow: Stateful tests
- [ ] <!---ci_include_integration--> Allow: Integration Tests
- [ ] <!---ci_include_performance--> Allow: Performance tests
- [ ] <!---ci_set_builds--> Allow: All Builds
- [ ] <!---batch_0_1--> Allow: batch 1, 2 for multi-batch jobs
- [ ] <!---batch_2_3--> Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ] <!---ci_exclude_style--> Exclude: Style check
- [x] <!---ci_exclude_fast--> Exclude: Fast test
- [x] <!---ci_exclude_asan--> Exclude: All with ASAN
- [x] <!---ci_exclude_tsan|msan|ubsan|coverage--> Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [x] <!---ci_exclude_aarch64|release|debug--> Exclude: All with aarch64, release, debug
---
- [ ] <!---ci_include_fuzzer--> Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ] <!---ci_exclude_ast--> Exclude: AST fuzzers
---
- [ ] <!---do_not_test--> Do not test
- [ ] <!---woolen_wolfdog--> Woolen Wolfdog
- [ ] <!---upload_all--> Upload binaries for special builds
- [x] <!---no_merge_commit--> Disable merge-commit
- [ ] <!---no_ci_cache--> Disable CI cache

pamarcos committed:
Use dummy launchers for compiler and linker when using clang-tidy

pamarcos committed:
Separate excludes and add some explanation

robot-ch-test-poll commented:
<!-- automatic status comment for PR #73647 from ClickHouse/ClickHouse:speed-up-build-tidy -->
*This is an automated comment for commit ea6d1ab906adcdbacf92ade816174776ff9385e8 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/73647/ea6d1ab906adcdbacf92ade816174776ff9385e8/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Style check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/73647/ea6d1ab906adcdbacf92ade816174776ff9385e8/style_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

pamarcos committed:
Add build_tidy job for sync PRs

pamarcos committed:
Improve the UX of the dummy launchers for tidy

pamarcos committed:
Fix typo

pamarcos committed:
Add linting error on purpose to test the CI

pamarcos committed:
Revert "Add linting error on purpose to test the CI"

This reverts commit 4b90e75b4d5c20a7bc58bee0fc84a32900b58a73.

pamarcos committed:
Ensure no real linkers are ever used

Turns out that if we set the dummy launchers at
the end of the CMakeLists.txt, that'll be the one
used for that project.

In order to make sure that the original ones are
used, we need to set it outside of the add_subdirectory
statement.

pamarcos committed:
Add EOF lines

divanik commented on cmake/utils.cmake:
```diff
@@ -121,3 +121,39 @@ function(get_target_property_list target outvar)
+# Set dummy linkers to avoid unnecessary work.
+macro(set_dummy_launchers_if_needed)
+    if(ENABLE_DUMMY_LAUNCHERS)
+        set(ORIGINAL_CMAKE_CXX_COMPILER_LAUNCHER ${CMAKE_CXX_COMPILER_LAUNCHER})
```
If set_dummy_launchers_if_needed is called twice we will overwrite ORIGINAL_CMAKE_C_COMPILER_LAUNCHER forever. Please, add a check to if that USING_DUMMY_LAUNCHERS is not set yet

pamarcos replied:
Yep, that's an excellent idea to prevent messing up with this in the future :+1:

divanik commented on cmake/utils.cmake:
```diff
@@ -121,3 +121,39 @@ function(get_target_property_list target outvar)
+
+# Some of our targets need to be linked for real. e.g. protoc, llvm-tlbgen and all its dependencies.
+macro(set_original_launchers_if_needed)
+    if(ENABLE_DUMMY_LAUNCHERS)
```
If set_original_launchers_if_needed is called before set_dummy_launchers_if_needed we will overwrite CMAKE_C_COMPILER_LAUNCHER with zero string. Add a check that USING_DUMMY_LAUNCHERS is already set

divanik commented on contrib/CMakeLists.txt:
```diff
@@ -67,17 +69,23 @@ add_contrib (brotli-cmake brotli)
 add_contrib (croaring-cmake croaring)
 add_contrib (zstd-cmake zstd)
+set_original_launchers_if_needed()
 add_contrib (zlib-ng-cmake zlib-ng)
```
Is it necessary for contribs to be added in a specific order or could we rearrange them to have only one section with original launcher?

pamarcos replied:
I asked myself the same, but didn't re-arrange them out of fear that something depended on something else. They shouldn't, but who knows. I'll give it a go, though.

pamarcos replied:
Well, surprise surprise... with my attempt to re-arrange all the ones that we need to compile with original linkers at the very beginning gave me a linking error in all builds. e.g. https://s3.amazonaws.com/clickhouse-builds/PRs/73647/10ffe578dd3d2cd7159fbf0cdd7614dc861d3390/package_release/build_log.log

```
Dec 23 18:59:58 [13101/13125] Linking CXX executable programs/clickhouse
Dec 23 18:59:58 FAILED: programs/clickhouse 
Dec 23 18:59:58 : && /usr/bin/clang++-19 --target=x86_64-linux-gnu --sysroot=/build/cmake/linux/../../contrib/sysroot/linux-x86_64/x86_64-linux-gnu/libc --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 -fdiagnostics-color=always -Xclang -fuse-ctor-homing -Wno-enum-constexpr-conversion -fsized-deallocation  -gdwarf-aranges -pipe -mssse3 -msse4.1 -msse4.2 -mpclmul -mpopcnt -fasynchronous-unwind-tables -ffile-prefix-map=/build=. -ftime-trace -falign-functions=32 -mbranches-within-32B-boundaries -ffp-contract=off  -fdiagnostics-absolute-paths -fstrict-vtable-pointers -Wall -Wextra -Wframe-larger-than=65536 -Weverything -Wpedantic -Wvla-cxx-extension -Wno-zero-length-array -Wno-c++98-compat-pedantic -Wno-c++98-compat -Wno-c++20-compat -Wno-sign-conversion -Wno-implicit-int-conversion -Wno-implicit-int-float-conversion -Wno-ctad-maybe-unsupported -Wno-disabled-macro-expansion -Wno-documentation-unknown-command -Wno-double-promotion -Wno-exit-time-destructors -Wno-float-equal -Wno-global-constructors -Wno-missing-prototypes -Wno-missing-variable-declarations -Wno-padded -Wno-switch-enum -Wno-undefined-func-template -Wno-unused-template -Wno-weak-template-vtables -Wno-weak-vtables -Wno-thread-safety-negative -Wno-enum-constexpr-conversion -Wno-unsafe-buffer-usage -Wno-switch-default -O2 -g -DNDEBUG -O3 -g  -flto=thin -fwhole-program-vtables -fno-pie --gcc-toolchain=/build/cmake/linux/../../contrib/sysroot/linux-x86_64 --ld-path=/usr/bin/ld.lld-19 -Wl,--no-export-dynamic -Wl,--gc-sections -Wl,--build-id=sha1 -Wl,--time-trace -no-pie -Wl,-no-pie -Wl,--allow-multiple-definition  -flto=thin -fwhole-program-vtables   -Xlinker --no-undefined @CMakeFiles/clickhouse.rsp -o programs/clickhouse && cd /build/build_docker/programs && /build/utils/check-style/check-large-objects.sh /build/build_docker && cd /build/build_docker/programs && mkdir -p /build/build_docker/programs/stripped/lib/debug/bin && mkdir -p /build/build_docker/programs/stripped/bin && cp clickhouse /build/build_docker/programs/stripped/bin/clickhouse && /usr/bin/llvm-objcopy-19 --only-keep-debug /build/build_docker/programs/stripped/bin/clickhouse /build/build_docker/programs/stripped/lib/debug/bin/clickhouse.debug && chmod 0644 /build/build_docker/programs/stripped/lib/debug/bin/clickhouse.debug && /usr/bin/llvm-strip-19 --strip-debug --remove-section=.comment --remove-section=.note /build/build_docker/programs/stripped/bin/clickhouse && /usr/bin/llvm-objcopy-19 --add-gnu-debuglink /build/build_docker/programs/stripped/lib/debug/bin/clickhouse.debug /build/build_docker/programs/stripped/bin/clickhouse
Dec 23 18:59:58 ld.lld-19: error: undefined symbol: grpc::SslServerCredentials(grpc::SslServerCredentialsOptions const&)
Dec 23 18:59:58 >>> referenced by GRPCServer.cpp:150 (./build_docker/./src/Server/GRPCServer.cpp:150)
Dec 23 18:59:58 >>>               src/clickhouse.lto.libdbms.a(GRPCServer.cpp.o at 2341790388).o:(DB::GRPCServer::start())
Dec 23 18:59:58 clang++-19: error: linker command failed with exit code 1 (use -v to see invocation)
Dec 23 18:59:58 ninja: build stopped: subcommand failed.
```

Of course, it happened because I added these ones at the very top. I need to add OpenSSL first, because otherwise it'll use the one from the system :upside_down_face:

divanik commented on docker/packager/packager:
```diff
@@ -353,10 +353,12 @@ def parse_env_variables(
         cmake_flags.append("-DENABLE_TESTS=1")
         cmake_flags.append("-DENABLE_EXAMPLES=1")
+        cmake_flags.append("-DENABLE_UTILS=1")
+        cmake_flags.append("-DENABLE_RUST=0")
```
Why do you need this line (about Rust)?

pamarcos committed:
Improvements following review's feedback

pamarcos committed:
Fix linking
